snakemake/executors.py

```diff
@@ -2168,3 +2171,878 @@ def run_wrapper(
+        lookup = {}
+        for zone in zones:
+            request = self._compute_cli.machineTypes().list(
+                project=self.project, zone=zone["name"]
+            )
+            lookup[zone["name"]] = self._retry_request(request)["items"]
+
+        # Only keep those that are shared, use last zone as a base
+        machine_types = {mt["name"]: mt for mt in lookup[zone["name"]]}
+        del lookup[zone["name"]]
+
+        # Update final list based on the remaining
+        to_remove = set()
+        for zone, types in lookup.items():
+            names = [x["name"] for x in types]
+            names = [name for name in names if "micro" not in name]
+            names = [name for name in names if not re.search("^(e2|m1)", name)]
+            for machine_type in list(machine_types.keys()):
+                if machine_type not in names:
+                    to_remove.add(machine_type)
+
+        for machine_type in to_remove:
+            del machine_types[machine_type]
+        return machine_types
+
+    def _add_gpu(self, gpu_count):
+        """Add a number of NVIDIA gpus to the current executor. This works
+           by way of adding nvidia_gpu to the job default resources, and also
+           changing the default machine type prefix to be n1, which is
+           the currently only supported instance type for using GPUs for LHS.
+        """
+        if not gpu_count or gpu_count == 0:
+            return
+
+        logger.debug(
+            "found resource request for {} GPUs. This will limit to n1 "
+            "instance types.".format(gpu_count)
+        )
+        self.workflow.default_resources.parsed["nvidia_gpu"] = gpu_count
+        self.workflow.default_resources.args.append("nvidia_gpu=%s" % gpu_count)
+        self._machine_type_prefix = self._machine_type_prefix or ""
+        if not self._machine_type_prefix.startswith("n1"):
+            self._machine_type_prefix = "n1"
+
+    def _generate_job_resources(self, job):
+        """given a particular job, generate the resources that it needs,
+           including default regions and the virtual machine configuration
+        """
+        # Right now, do a best effort mapping of resources to instance types
+        cores = job.resources.get("_cores", 1)
+        mem_mb = job.resources.get("mem_mb", 15360)
+        disk_mb = job.resources.get("disk_mb", 128000)
+
+        # Convert mb to gb, add buffer of 50
+        disk_gb = math.ceil(disk_mb / 1024) + 10
+
+        # Look for if the user wants an nvidia gpu
+        gpu_count = job.resources.get("nvidia_gpu") or job.resources.get("gpu")
+        gpu_model = job.resources.get("gpu_model")
+
+        # If a gpu model is specified without a count, we assume 1
+        if gpu_model and not gpu_count:
+            gpu_count = 1
+
+        # Update default resources using decided memory and disk
+        self.workflow.default_resources = self.default_resources
```

Comment:
Default resources should never be modified at this stage.

Reply:
This was done because something wasn't working, I really don't want to go through debugging it again because I know it's there for a reason. :(